Store `display: none` RenderStyle objects

graouts opened the pull request:
#### 519c7ed4ac755138208e8daf71057c4ac8a2f810
<pre>
Store `display: none` RenderStyle objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=267683">https://bugs.webkit.org/show_bug.cgi?id=267683</a>

Reviewed by Antti Koivisto.

We currently store `display: contents` RenderStyle objects on Element. In order to implement support
for animating the `display` property, we need to also store styles in the `display: none` case so that
we may return such stored styles in `Element::existingComputedStyle()`.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasDisplayContents const):
(WebCore::Element::hasDisplayNone const):
(WebCore::Element::storeDisplayContentsOrNoneStyle):
(WebCore::Element::clearDisplayContentsOrNoneStyle):
(WebCore::Element::existingComputedStyle const):
(WebCore::Element::renderOrDisplayContentsStyle const):
(WebCore::Element::resetComputedStyle):
(WebCore::Element::storeDisplayContentsStyle): Deleted.
(WebCore::Element::clearDisplayContentsStyle): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::displayContentsOrNoneStyle const):
(WebCore::ElementRareData::setDisplayContentsOrNoneStyle):
(WebCore::ElementRareData::useTypes const):
(WebCore::ElementRareData::displayContentsStyle const): Deleted.
(WebCore::ElementRareData::setDisplayContentsStyle): Deleted.
* Source/WebCore/dom/Node.cpp:
(WebCore::stringForRareDataUseType):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updatePseudoElement):

Canonical link: <a href="https://commits.webkit.org/273288@main">https://commits.webkit.org/273288@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d54830179729c5618edab6f5ca75a579fe55ab3b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34785 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13606 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36793 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37472 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31390 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16068 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10691 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30355 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11578 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30976 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10072 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38737 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31375 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36178 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10249 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8153 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34154 "Found 1 new API test failure: /WebKitGTK/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12077 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10795 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4499 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11143 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -98,6 +98,9 @@ class ElementRareData : public NodeRareData {
     void setDisplayContentsStyle(std::unique_ptr<RenderStyle> style) { m_displayContentsStyle = WTFMove(style); }
 
+    RenderStyle* displayNoneStyle() const { return m_displayNoneStyle.get(); }
+    void setDisplayNoneStyle(std::unique_ptr<RenderStyle> style) { m_displayNoneStyle = WTFMove(style); }
```
`display:none` and `display:contents` are mutually exclusive. You don't need a separate field, you can just rename `m_displayContentsStyle` to `m_displayNoneOrContentsStyle` (along with the accessor functions).

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2466,32 +2466,40 @@ void Element::invalidateEventListenerRegions()
 }
 
-bool Element::hasDisplayContents() const
+bool Element::hasDisplayType(DisplayType displayType) const
```
I realize this is private but it should be clear in its name it only works for stored styles, like hasDisplayForStoredStyle(). Not sure a helper is even needed really.

graouts replied:
I was adding it to avoid some code duplication, but it's also not exactly rocket science, so I might remove it altogether.

graouts committed:
Store `display: none` RenderStyle objects
https://bugs.webkit.org/show_bug.cgi?id=267683

Reviewed by Antti Koivisto.

We currently store `display: contents` RenderStyle objects on Element. In order to implement support
for animating the `display` property, we need to also store styles in the `display: none` case so that
we may return such stored styles in `Element::existingComputedStyle()`.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasDisplayContents const):
(WebCore::Element::hasDisplayNone const):
(WebCore::Element::storeDisplayContentsOrNoneStyle):
(WebCore::Element::clearDisplayContentsOrNoneStyle):
(WebCore::Element::existingComputedStyle const):
(WebCore::Element::renderOrDisplayContentsStyle const):
(WebCore::Element::resetComputedStyle):
(WebCore::Element::storeDisplayContentsStyle): Deleted.
(WebCore::Element::clearDisplayContentsStyle): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::displayContentsOrNoneStyle const):
(WebCore::ElementRareData::setDisplayContentsOrNoneStyle):
(WebCore::ElementRareData::useTypes const):
(WebCore::ElementRareData::displayContentsStyle const): Deleted.
(WebCore::ElementRareData::setDisplayContentsStyle): Deleted.
* Source/WebCore/dom/Node.cpp:
(WebCore::stringForRareDataUseType):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updatePseudoElement):

Canonical link: https://commits.webkit.org/273288@main
